### ISSUE TITLE
Improve evaluation of memory usage in testInitState

### ISSUE THREAD

**Issue by tkuchida** [opensim-org/opensim-core#1263]
See [testInitState.cpp L156](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Simulation/Test/testInitState.cpp#L156).

### Issue

If memory is freed behind the scenes immediately after L156, then `mem0` will not represent the "base footprint" as assumed and the evaluation of memory usage will be ineffective (currently, if `current_size < mem0`, we set `model_size = 0` and leaks could slip through). We also don't want to throw an exception if `current_size < mem0`, as we already have lots of false alarms on CI (e.g., testContactGeometry).

### Suggestions

1. Perhaps the entire operation of setting `mem0`, loading a model, and computing the memory increase should be repeated until a nonzero memory increase is detected.

2. If the model is small, perhaps it can always fit within the memory already allocated (I don't know the details on every platform). I suggest loading *n* copies of the model and then dividing the memory increase by *n*. Perhaps use the following algorithm:

```
n <- 1
while (1) {
    mem0 <- current memory footprint
    loadNCopiesOfModel(n)
    mem1 <- current memory footprint
    if (mem1 > mem0)
        break
    ++n
}
memoryIncrease <- (mem1-mem0) / n
```

**Comment by tkuchida** [comment]
An analogous issue exists in testComponents.cpp, which would benefit from an analogous solution.

**Comment by aseth1** [comment]
Both seem reasonable to me, but 2. does not address the problem if `mem0` is inaccurate.  In which case we will loop N times and remain unsuccessful. This is what is currently happening in #1205.

> We also don't want to throw an exception if current_size < mem0, as we already have lots of false alarms on CI (e.g., testContactGeometry).

What if you have looped through N times and it still evaluates to less than mem0? 

We could bake in the memory estimate of a model based on the mean (or median) value (over N samples) by adding a utility auxiliary testing so it could be used by other test. The median value would be more accurate if a few of the evaluations are completely wrong (like 0 or even negative). We could also evaluate mem0 in a similar way,  estimating over N calls with a brief/mandatory wait. 

@tkuchida are you tackling this now? Otherwise, I will do it.

**Comment by tkuchida** [comment]
> but 2. does not address the problem if `mem0` is inaccurate.  In which case we will loop N times and remain unsuccessful.

As described by the pseudocode, each iteration of the algorithm begins by establishing a "base footprint" and then loads *n* copies of the model, where *n* increases with each iteration. Unless huge chunks of memory are being freed immediately after `mem0` is stored on *every* iteration, then I believe this algorithm may work—though I have not tried it.

> What if you have looped through N times and it still evaluates to less than mem0?

As *n* increases, I would expect this scenario to be less likely—unless the memory freeing mechanism is exceptionally vigilant.

> We could bake in the memory estimate of a model based on the mean (or median) value (over N samples) by adding a utility auxiliary testing so it could be used by other test.

I don't know the end of this sentence means, but I don't think using a mean or median is correct. Using a mean implies that each measurement is believed to be equally trustworthy, when we know that if `mem1 < mem0` then we should throw out that data point. I think we just want the max over all iterations of `(mem1 > mem0) ? mem1-mem0 : 0`.

**Comment by aseth1** [comment]
> As described by the pseudocode

Right .. the `<-` being the assignment of a value - I misread that.

> As n increases, I would expect this scenario to be less likely—unless the memory freeing mechanism is exceptionally vigilant

But you still throw an exception, yes? 

> I don't know the end of this sentence means ...

 I meant put the functions that estimate of memory usage based on N samples in the `auxiliaryTestFunctions` header instead of the test cases. Then multiple tests like `testComponents` could use the same utilities.

I agree the mean is incorrect. Perhaps max is the right way to go unless their is anomaly that makes one measure huge and now you are throwing away N samples of information. Why not use the median?

> when we know that if mem1 < mem0 then we should throw out that data point

Which data point? mem1 or mem0? Or the entire case?

Anyhow, I will try something in the next day or so.

**Comment by tkuchida** [comment]
> But you still throw an exception, yes?

Yes, I think the test should fail if the memory-leak detector is malfunctioning.

> I meant put the functions that estimate of memory usage based on N samples in the auxiliaryTestFunctions header instead of the test cases.

:+1:

> I agree the mean is incorrect. Perhaps max is the right way to go unless their is anomaly that makes one measure huge and now you are throwing away N samples of information. Why not use the median?

Without a proposed algorithm, I don't know whether the median is best. If you want at least one positive memory footprint increase, then the loop should just exit on `mem1 > mem0`. If you want exactly `s` samples of memory footprint increases, then the median seems fine. Collecting `s` samples and then taking the median is wrong because most of the `s` samples could be zero. In short, I think collecting `s > 1` samples where the memory footprint strictly increased and then taking the median sounds pretty solid.

> Which data point? mem1 or mem0? Or the entire case?

I would discard the entire case.

**Comment by aseth1** [comment]
addressed by #1267
